crypto/fipsmodule/ec/ec_nistp.c

```diff
@@ -365,68 +368,77 @@ static inline void select_point_from_table(const ec_nistp_meth *ctx,
+//
+// The function has four modes of operation: base and arbitrary, and their
+// non constant-time variants base-non-ct and arbitrary-non-ct. The base modes
+// are used for scalar multiplication with the base point, and the arbitrary
+// modes for an arbitrary point. The non-ct variants are used when computing
+// the so-called public multiplication where the scalar is not secret so the
+// operation doesn't have to be implemented in constant-time.
+enum scalar_mul_core_mode { base, base_nonct, arbitrary, arbitrary_nonct };
+//
+// In arbitrary modes the input table is a single table with odd multiples of P:
+//     [2i + 1]P for i in [0, SCALAR_MUL_TABLE_NUM_POINTS - 1],
+// In base modes the input table is a full table of m single subtables:
+//     [(2j + 1) * 2^(i*w)]P for i in [0, m - 1] and
+//                           for j in [0, SCALAR_MUL_TABLE_NUM_POINTS - 1].
 //
 // The required product, [scalar]P, is computed by the following algorithm.
 //     1. Initialize the accumulator with the point from |table|
 //        corresponding to the most significant digit s_{m-1} of the scalar.
 //     2. For digits s_i starting from s_{m-2} down to s_0:
-//     3.   Double the accumulator w times. (note that doubling a point [a]P
-//          w times results in [2^w*a]P).
+//     3.   Double the accumulator w times if in an arbitrary mode.
+//          (note that doubling a point [a]P w times results in [2^w*a]P).
 //     4.   Read from |table| the point corresponding to abs(s_i),
 //          negate it if s_i is negative, and add it to the accumulator.
 //     5. Subtract P from the result if the scalar is even.
-//
-// Note: this function is constant-time.
-void ec_nistp_scalar_mul(const ec_nistp_meth *ctx,
-                         ec_nistp_felem_limb *x_out,
-                         ec_nistp_felem_limb *y_out,
-                         ec_nistp_felem_limb *z_out,
-                         const ec_nistp_felem_limb *x_in,
-                         const ec_nistp_felem_limb *y_in,
-                         const ec_nistp_felem_limb *z_in,
-                         const EC_SCALAR *scalar) {
-  // Make sure that the max table size is large enough.
-  assert(SCALAR_MUL_TABLE_MAX_NUM_FELEM_LIMBS >=
-         SCALAR_MUL_TABLE_NUM_POINTS * ctx->felem_num_limbs * 3);
-
-  // Table of multiples of P = (x_in, y_in, z_in).
-  ec_nistp_felem_limb table[SCALAR_MUL_TABLE_MAX_NUM_FELEM_LIMBS];
-  generate_table(ctx, table, x_in, y_in, z_in);
+static inline void scalar_mul_core(const ec_nistp_meth *ctx,
+                                   ec_nistp_felem_limb *x_out,
+                                   ec_nistp_felem_limb *y_out,
+                                   ec_nistp_felem_limb *z_out,
+                                   const EC_SCALAR *scalar,
+                                   const ec_nistp_felem_limb *table,
+                                   enum scalar_mul_core_mode mode) {
+  // Mode based invariants.
```

Comment:
Just something to ponder later. Is something else using `scalar_mul_core` apart from the three function further below? If not, while these modes attempts to abstract some semantics, it's almost too much abstraction. If the set of functions that invoke, and will ever invoke `scalar_mul_core`, one could have this as three arguments.

It also appears there is some implicit relationship between the table argument and the mode argument. Although, any mistakes mis-configuring that e.g. use `mode != base` for a full table, will likely cause the unit tests to have severe hiccups.